### ISSUE TITLE
Compare (diff) two files

### ISSUE THREAD

**Issue by krstp** [zed-industries/zed#17100]
### Check for existing issues

- [X] Completed

### Describe the feature

I would like to have an ability to compare/diff two files side by side. I see plenty of requests for `git diff` but I do not see regular diff request anwhere.

### If applicable, add mockups / screenshots to help present your vision of the feature

Side by side two files diff ability.

**Comment by Moshyfawn** [comment]
I think #4523 was originally what you were suggesting here, but it got "commented into being a `git diff view`", so I'll leave this one open.

**Comment by Joride** [comment]
Any update on this (or https://github.com/zed-industries/zed/issues/4523)? 

Currently I need to switch to a different editor to view side-by-side the change I made before committing. This is quite the nuisance (especially since Zed's creates new files with "~filename", which sometimes causes me to edit a file I was not even aware of existed and is the wrong one :) . In short: would be great to have a diff view in Zed. (Much more important to me then any 'AI' feature, honestly).

**Comment by raul-gherman** [comment]
jumping back to any other editor to diff 2 arbitrary files is breaking my back

would love to have it here -- basic features should be in place before `fancy` ones -- and this is somehow basic

**Comment by raul-gherman** [comment]
> Any update on this (or [#4523](https://github.com/zed-industries/zed/issues/4523))?
> 
> Currently I need to switch to a different editor to view side-by-side the change I made before committing. This is quite the nuisance (especially since Zed's creates new files with "~filename", which sometimes causes me to edit a file I was not even aware of existed and is the wrong one :) . In short: would be great to have a diff view in Zed. (Much more important to me then any 'AI' feature, honestly).

last update 1 month ago

**Comment by krstp** [comment]
I've recently found myself needing diffing features, which has led me to return to Sublime Text as it excels in this area.

Several useful features are currently missing in Zed:

* Ability to diff two files in open tabs/buffers without requiring users to save the files first
* Full git diff functionality between independent commits/branches - while there appears to be some git diff implementation, it only shows current changes rather than providing comprehensive diff functionality
* Support for shell piping, allowing Zed to be used within command pipelines

Sublime Text handles all these capabilities flawlessly, particularly the ability to pipe `git diff` output directly to the editor for visualization.

**Comment by alex-milanovic** [comment]
Any update on this? This is one of the main reasons I'm not fully switching from PyCharm

**Comment by ivanglushko** [comment]
would love that for now gotta use external stuff

**Comment by ralgozino** [comment]
Folks, please stop commenting "me too" on this closed issue. If you take a moment to read you'll see that it has been closed as a duplicate of #4523, so go at put a 👍 over there.

I'm sure the Zed team already understood the request and they'll work on it when they can.

**Comment by bpresles** [comment]
Reopen this issue or #4523 as what was implemented is git diff, not diff between two arbitrary files. Diff between arbitrary files means comparing files even when they don't differ from what's on the git repository remote branch.

**Comment by ralgozino** [comment]
@bpresles in May 14th #4523 was reopened saying exactly the same as you are saying.

Then on June 18th it was closed because the feature to see the diff between two arbitrary files was merged in the linked PR #32922

You can diff two files with Zed with the following command in the terminal:

```bash
zed --diff file1 file2
```

I don't know if there's a way of doing it directly from the UI.

**Comment by krstp** [comment]
"zed diff" is no different that regular "diff", I mean what is the benefit of having a feature and running it from the command line? ¯\_(ツ)_/¯ 

The core request is to be able to compare two open tabs against each other, so driven from UI as Zed is an UI.

So yes, I vote for reopening as well.

**Comment by ralgozino** [comment]
> "zed diff" is no different that regular "diff", I mean what is the benefit of having a feature and running it from the command line? ¯_(ツ)_/¯
> 
> The core request is to be able to compare two open tabs against each other, so driven from UI as Zed is an UI.
> 
> So yes, I vote for reopening as well.

Have you tested it? the command opens Zed and shows the diff using Zed's UI. It is not like regular "diff".

I agree that it would be nice(r) to have a menu to select the files from the file explorer or choosing open tabs. But in the meantime you can do that with the command.

**Comment by bpresles** [comment]
OK I'll reopen a new feature request specifying clearly it's about using UI for both selecting files and diff

**Comment by mrsarm** [comment]
There is no way to compare side by side 2 files in Zed, just support to open "diff" content like a ".diff" file, or see changes against the git repo as a diff content, so this issue does not duplicate the other ticket where what is requested here was not implemented in anyway.

**Comment by krstp** [comment]
@JosephTLyons Any chance we can reopen this ticket? Thank you for the attention and making Zed great the way it is.

**Comment by krstp** [comment]
@JosephTLyons  Nudging again here as there does not seem to be resolution to it yet and https://github.com/zed-industries/zed/pull/35255 does not solve the core request described here.

Running now on latest preview: `Zed Preview 0.206.0` and the functionality is not here.

I hope we can reopen and provide the requested functionality. Thanks.

**Comment by krstp** [comment]
Well correction. Now we have `Compare marked files`, but it requires usage of UI, and there is not way compare tab-to-tab option. Also the keyboard shortcut `Option+D` is not working. Great to see this feature in place!

**Comment by robert-zaremba** [comment]
Let's reopen this issue. There is not a proper mode to make a diff of 2 file and merge them.
NOTE:  this is not the same as git diff / git merge.
A great, powerful example of such mode is Emacs' Ediff. See this [video](https://www.youtube.com/watch?v=pSvsAutseO0) how powerful it is. Check also [Kompare](https://apps.kde.org/kompare/) (from KDE suite) for a good example. Given that Zed already has git diff interface, this should be fairly simple.
This is one of the few features that makes me constantly moving to Emacs.

**Comment by techdad** [comment]
This is a blocker for me to use Zed.
I use the compare files feature in VScode (that are NOT in git) as well as the compare open to saved on a daily basis.
Pity. Zed looked so promising in all other ways

**Comment by Van0SS** [comment]
Upvoting on this one, meanwhile going back to vscode

**Comment by SomeoneToIgnore** [comment]
<img width="1728" height="1083" alt="Image" src="https://github.com/user-attachments/assets/42eeac41-0e84-43b8-811f-b69a4b6ce005" />

This seems to be implemented?

**Comment by Van0SS** [comment]
Nice! But I guess was looking at similar to vscode "Compare active file with clipboard" functionality, when I don't have permanent files, just some copy paste in my clipboard

**Comment by SomeoneToIgnore** [comment]
<img width="1728" height="285" alt="Image" src="https://github.com/user-attachments/assets/75f40518-5b00-4d43-9a80-b5120a460a37" />

We really miss the docs on all these, as those were added as external contributions in different time...
That would be a nice PR do create.

**Comment by Van0SS** [comment]
Wow! You found it! Thanks!

Now I need side by side comparison haha, but I guess I need upvote here:
https://github.com/zed-industries/zed/discussions/26770

Also would be nice to have highlighted different words, not just whole line/paragraph, but not even sure how to formulate it -_-

zed:
<img width="429" height="89" alt="Image" src="https://github.com/user-attachments/assets/8e8a9f4b-a2c1-4fc3-9212-4a59e545d0bd" />

vscode:
<img width="1040" height="61" alt="Image" src="https://github.com/user-attachments/assets/58f9ab13-b203-42fc-a15a-b356e6ad7c39" />

**Comment by krstp** [comment]
One thing should be said, what appears to me is Zed having some issue with temporary/ephemeral/not-saved buffers. I say it based of:
- lack of feature even by means of command panel to compare two tabs with not-saved buffers;
- lack of formatting/prettying of JSONs unless it is saved.
- ruff formatting on unsaved buffers also seems to be an issue: https://github.com/zed-industries/zed/issues/17098#issuecomment-3571844243

Hard for me to say now if non-saved buffers are the exact issue, but based on above use cases not implemented yet, that seems the obvious candidate.

Second fold is... the same issue/request coming back in the GH issues every now and then, so this is somewhat telling. Either VS or Sublime or other editors seem to have this feature, hence people returning here with the same request as current implementation as much as it is useful it is not necessarily intuitive based on other market available products.

Either way, great to have Zed as it is right now!

**Comment by Sillocan** [comment]
Ahhh I kept searching for "Compare" and kept finding no reults. Thanks For posting a screenshot: "diff" is the magic phrase!
